File: lib/src/main/java/arsatech/co/utils/SearchAlgorithms.java
package arsatech.co.utils;

import android.util.Pair;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Created by Aref Bahreini Nejad on 27/04/2019.
 * Updated on 27/04/2019
 */
public class SearchAlgorithms {

	private static String normalize(String str) {
		return StringUtils.strToPersian(str.toLowerCase());
	}

	public static class InStrings {

		public static ArrayList<String> startingWithSearch(ArrayList<String> list, String query) {
			query = normalize(query);
			ArrayList<String> result = new ArrayList<>();
			for (int i = 0; i < list.size(); i++) {
				String temp = normalize(list.get(i));
				;
				if (temp.startsWith(query)) {
					result.add(list.get(i));
				}
			}
			return result;
		}

		public static ArrayList<String> exactSearch(ArrayList<String> list, String query) {
			query = normalize(query);
			ArrayList<String> result = new ArrayList<>();
			for (int i = 0; i < list.size(); i++) {
				String temp = normalize(list.get(i));
				;
				if (temp.equals(query)) {
					result.add(list.get(i));
				}
			}
			return result;
		}

		public static ArrayList<String> containSearch(ArrayList<String> list, String query) {
			query = normalize(query);
			ArrayList<String> result = new ArrayList<>();
			for (int i = 0; i < list.size(); i++) {
				int lastCharIndex;
				String temp = normalize(list.get(i));
				;
				boolean isOK = true;
				for (int j = 0; j < query.length(); j++) {
					if (!temp.contains((query.charAt(j) + ""))) {
						isOK = false;
						break;
					}
					lastCharIndex = temp.indexOf(query.charAt(j));
					if (temp.length() > lastCharIndex + 1) {
						temp = temp.substring(lastCharIndex + 1);
					}
				}
				if (isOK) {
					result.add(list.get(i));
				}
			}
			return result;
		}

		public static ArrayList<String> exactContainSearch(ArrayList<String> list, String query) {
			query = normalize(query);
			ArrayList<String> result = new ArrayList<>();
			for (int i = 0; i < list.size(); i++) {
				String temp = normalize(list.get(i));
				;
				if (temp.contains(query)) {
					result.add(temp);
				}
			}
			return result;
		}

		public static ArrayList<String> faultyContainSearch(ArrayList<String> list, String query) {
			query = normalize(query);
			ArrayList<Pair<String, Float>> result = new ArrayList<>();
			for (int i = 0; i < list.size(); i++) {
				String temp = normalize(list.get(i));
				;
				if (temp.charAt(0) != query.charAt(0)) {
					continue;
				}
				float score = 1;
				int lastCharIndex;
				for (int j = 1; j < query.length(); j++) {
					if (!temp.contains(query.charAt(j) + "")) {
						continue;
					}
					score++;
					lastCharIndex = temp.indexOf(query.charAt(j));
					if (temp.length() > lastCharIndex + 1) {
						temp = temp.substring(lastCharIndex + 1);
					}
				}
				score /= query.length();
				result.add(new Pair<>(list.get(i), score));
			}
			Collections.sort(result, (o1, o2) -> {
				if (o1.second < o2.second) {
					return 1;
				} else if (o1.second > o2.second) {
					return -1;
				}
				return 0;
			});
			ArrayList<String> finalResult = new ArrayList<>();
			for (int i = 0; i < result.size(); i++) {
				if (result.get(i).second > 0.4) {
					finalResult.add(result.get(i).first);
				}
			}
			return finalResult;
		}

	}

}

File: lib/src/main/kotlin/com/arefdev/base/utils/SearchAlgorithms.kt
package com.arefdev.base.utils

import android.util.Pair

/**
 * Updated on 21/09/2023
 *
 * @author <a href="https://github.com/arefbhrn">Aref Bahreini</a>
 */
object SearchAlgorithms {

    private fun normalize(str: String): String {
        return StringUtils.strToPersian(str.lowercase())
    }

    object InStrings {

        fun startingWithSearch(list: List<String>, query: String): List<String> {
            val mQuery = normalize(query)
            val result = mutableListOf<String>()
            for (i in list.indices) {
                val temp = normalize(list[i])
                if (temp.startsWith(mQuery)) {
                    result.add(list[i])
                }
            }
            return result
        }

        fun exactSearch(list: List<String>, query: String): List<String> {
            val mQuery = normalize(query)
            return list.filter { normalize(it) == mQuery }
        }

        fun containSearch(list: List<String>, query: String): List<String> {
            val mQuery = normalize(query)
            return list.filter {
                var normalizedItem = normalize(it)
                var isOK = true
                for (j in mQuery.indices) {
                    val lastCharIndex = normalizedItem.indexOf(mQuery[j])
                    if (lastCharIndex < 0) {
                        isOK = false
                        break
                    }
                    if (normalizedItem.length > lastCharIndex + 1) {
                        normalizedItem = normalizedItem.substring(lastCharIndex + 1)
                    }
                }
                return@filter isOK
            }
        }

        fun exactContainSearch(list: List<String>, query: String): List<String> {
            val mQuery = normalize(query)
            return list.filter { normalize(it).contains(mQuery) }
        }

        fun faultyContainSearch(list: List<String>, query: String): List<String> {
            val mQuery = normalize(query)
            val result = mutableListOf<Pair<String, Float>>()
            for (item in list) {
                var normalizedItem = normalize(item)
                if (normalizedItem[0] != mQuery[0]) {
                    continue
                }
                var score = 1f
                for (j in 1 until mQuery.length) {
                    val lastCharIndex = normalizedItem.indexOf(mQuery[j])
                    if (lastCharIndex < 0) {
                        continue
                    }
                    score++
                    if (normalizedItem.length > lastCharIndex + 1) {
                        normalizedItem = normalizedItem.substring(lastCharIndex + 1)
                    }
                }
                score /= mQuery.length.toFloat()
                result.add(Pair(item, score))
            }
            result.sortWith { a, b ->
                return@sortWith if (a.second < b.second)
                    1
                else if (a.second > b.second)
                    -1
                else
                    0
            }
            val finalResult = mutableListOf<String>()
            for (item in result) {
                if (item.second > 0.4) {
                    finalResult.add(item.first)
                }
            }
            return finalResult
        }
    }
}
